File: scripts/release/translate-notes.py
#!/usr/bin/env python3

from __future__ import annotations

import json
import os
import sys
import urllib.error
import urllib.request


def translate_with_openai_api(text: str, api_key: str, model: str) -> str:
    payload = {
        "model": model,
        "messages": [
            {
                "role": "system",
                "content": "Translate GitHub release notes from English to Simplified Chinese. Keep markdown structure, code blocks, links, issue numbers, and version identifiers unchanged.",
            },
            {"role": "user", "content": text},
        ],
        "temperature": 0.2,
    }

    req = urllib.request.Request(
        "https://api.openai.com/v1/chat/completions",
        data=json.dumps(payload).encode("utf-8"),
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )

    with urllib.request.urlopen(req, timeout=60) as resp:
        body = resp.read().decode("utf-8")
        data = json.loads(body)
        return data["choices"][0]["message"]["content"].strip()


def main() -> int:
    if len(sys.argv) != 3:
        print("Usage: python scripts/release/translate-notes.py <english-input.md> <chinese-output.md>", file=sys.stderr)
        return 1

    input_path = sys.argv[1]
    output_path = sys.argv[2]

    with open(input_path, "r", encoding="utf-8") as f:
        english_body = f.read().strip()

    if not english_body:
        with open(output_path, "w", encoding="utf-8") as f:
            f.write("")
        return 0

    api_key = os.environ.get("OPENAI_API_KEY", "").strip()
    model = os.environ.get("RELEASE_TRANSLATION_MODEL", "gpt-4o-mini").strip() or "gpt-4o-mini"

    if not api_key:
        with open(output_path, "w", encoding="utf-8") as f:
            f.write(english_body)
        print("OPENAI_API_KEY not set. Fallback to English notes.")
        return 0

    try:
        translated = translate_with_openai_api(english_body, api_key, model)
    except (urllib.error.URLError, urllib.error.HTTPError, KeyError, IndexError, json.JSONDecodeError) as exc:
        with open(output_path, "w", encoding="utf-8") as f:
            f.write(english_body)
        print(f"Translation failed ({exc}). Fallback to English notes.")
        return 0

    with open(output_path, "w", encoding="utf-8") as f:
        f.write(translated)

    print(f"Translated notes written to {output_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/release/translate-notes.ts
import { readFileSync, writeFileSync } from 'node:fs';

export function resolveChatCompletionsEndpoint(baseUrl: string): string {
  const trimmed = baseUrl.trim();
  if (!trimmed) {
    return 'https://api.openai.com/v1/chat/completions';
  }

  const normalized = trimmed.replace(/\/+$/, '');
  if (normalized.endsWith('/chat/completions')) {
    return normalized;
  }

  return `${normalized}/chat/completions`;
}

type ChatCompletionsResponse = {
  choices?: Array<{
    message?: {
      content?: string;
    };
  }>;
};

async function translateWithApi(params: {
  text: string;
  apiKey: string;
  model: string;
  endpoint: string;
}): Promise<string> {
  const { text, apiKey, model, endpoint } = params;

  const payload = {
    model,
    messages: [
      {
        role: 'system',
        content:
          'Translate GitHub release notes from English to Simplified Chinese. Keep markdown structure, code blocks, links, issue numbers, and version identifiers unchanged.',
      },
      {
        role: 'user',
        content: text,
      },
    ],
    temperature: 0.2,
  };

  const response = await fetch(endpoint, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    throw new Error(`Translation request failed with status ${response.status}`);
  }

  const data = (await response.json()) as ChatCompletionsResponse;
  const translated = data.choices?.[0]?.message?.content?.trim();

  if (!translated) {
    throw new Error('Translation response did not include content');
  }

  return translated;
}

async function main(): Promise<void> {
  const [inputPath, outputPath] = process.argv.slice(2);

  if (!inputPath || !outputPath) {
    console.error('Usage: bun run scripts/release/translate-notes.ts <english-input.md> <chinese-output.md>');
    process.exit(1);
  }

  const englishBody = readFileSync(inputPath, 'utf-8').trim();
  if (!englishBody) {
    writeFileSync(outputPath, '', 'utf-8');
    return;
  }

  const apiKey = (process.env.OPENAI_API_KEY || '').trim();
  const model = (process.env.RELEASE_TRANSLATION_MODEL || 'gpt-4o-mini').trim() || 'gpt-4o-mini';
  const endpoint = resolveChatCompletionsEndpoint(process.env.RELEASE_TRANSLATION_BASE_URL || '');

  if (!apiKey) {
    writeFileSync(outputPath, englishBody, 'utf-8');
    console.log('OPENAI_API_KEY not set. Fallback to English notes.');
    return;
  }

  try {
    const translated = await translateWithApi({
      text: englishBody,
      apiKey,
      model,
      endpoint,
    });
    writeFileSync(outputPath, translated, 'utf-8');
    console.log(`Translated notes written to ${outputPath}`);
  } catch (error) {
    writeFileSync(outputPath, englishBody, 'utf-8');
    console.log(`Translation failed (${String(error)}). Fallback to English notes.`);
  }
}

if (import.meta.main) {
  await main();
}
